### DISCUSSION TITLE
Max string length 256 bytes

### DISCUSSION THREAD

**Discussion by jkotlinski** [jkotlinski/durexforth#461] [Ideas]
I am (again) toying with the idea to make max string length 256 bytes.

Good: Saves one byte per string definition.
Bad: MML definitions will have to move to files.

What do you think?

**Comment by lonetech** [comment]
AIUI, the reason it's 2 bytes is because `lits` reads a word length. Other schemes are possible, e.g. read a byte, and if it has the high bit set, use it as MSB (with the high bit cleared) and read another for LSB. This would reduce max string length from 64k (impossible for other reasons) to 32k (also larger than needed), while allowing the size reductions for strings under 128 bytes. But that would be more complex logic, thus bigger and slower `s"` and `lits`. The latter could be reduced by having separate entry points for the two variants. 

The 8-bit count is a common and well defined format (counted string), and short strings are the more common case. It wouldn't be totally unreasonable to let compilation of longer strings perform an extra move as data overflows from 8-bit to 16-bit length, or even vice versa. But the flexibility of such a dynamic length `s"` has a cost too, and it may outweigh current uses. 

It looks to me like the base system has almost no uses of `lits`, once in ls and debug and twice in v (it's used through `."` or `s"`). Thus the savings per definition wouldn't immediately approach the cost of supporting two variants. On the other hand, the MML demo is the only thing using long strings, and it could just include words to do so. 

The standard [parsing](https://forth-standard.org/standard/usage#usage:parsing) section does allow a limit to the counted string length. So I think reducing these words to 8-bit lengths is compliant. Shorter than the counted length isn't permitted, though, so given the size of our input buffers we still need the refill ability of getc. 

I think it would make sense given common uses to have the core `s"` limited to 8-bit, but length 0-255, making `lits` read a counted string from the return address. The long `s"` and `lits` could simply be moved into mmldemo, avoiding the need for separate file reading. Notably mmldemo does use an empty string.

**Comment by jkotlinski** [comment]
Going back in time, originally only 256 bytes string length was allowed. This was expanded to 65536 bytes for the MML support.

In hindsight, I think this was the wrong thing to do. 256 is normal for 8-bit Forths, and supported by the standard.

**Reply by Whammo** [reply]
It is definitely cool though. I'd used MML in game development and was pleased.
